doc/man1/roarclientpass.1
.\" roarclientpass.1:

.TH "roarclientpass" "1" "July 2010" "RoarAudio" "System User's Manual: roarclientpass"

.SH NAME

roarclientpass \- Passing clients to and creating new listen sockets on RoarAudio sound server

.SH SYNOPSIS

roarclientpass [OPTIONS]...

.SH DESCRIPTION

This program let you pass new clients to a RoarAudio sound server as as a proxy and
let you add new listen sockets to a running server.

.SH "OPTIONS"

.TP
\fB--server    SERVER\fR
Set server hostname.

.TP
\fB--stdin\fR
Client is on stdin.

.TP
\fB--stdout\fR
Client is on stdout.

.TP
\fB--stdio\fR
Same as \-\-stdin \-\-stdout.

This option may for example be used in case of being runed from a inetd.

.TP
\fB--stderr\fR
Client is on stderr.

Beside to make stderr known as client fh it will disable error output on stderr.
This \fBmust\fR be used in case this program is used from inetd or simular tools.

.TP
\fB--client-fh FH\fR
Client is on FH.

This is basicly the same as \-\-client-fh on roard(1).

.TP
\fB--proto PROTO\fR
Set the protocol used by the client.

Defaults to RoarAudio.

.TP
\fB--byteorder BO\fR
Sets the byteorder used by the client.

Defaults to network byte order.

.TP
\fB--listen\fR
Client is a listen connection.

This is not the same as \-\-mode listen.

.TP
\fB--mode MODE\fR
Set mode of operation: none, listen or connect

Defaults to none.

.TP
\fB--bind BIND\fR
Set host/node/path for mode listen and connect.

.TP
\fB--port PORT\fR
Set port for mode listen and connect.

.TP
\fB--help\fR
Show this help

.SH "SEE ALSO"
\fBroartips\fR(7),
\fBlibroar\fR(7),
\fBRoarAudio\fR(7).

.SH "HISTORY"

For history information see \fBRoarAudio\fR(7).

.\" ll


ChangeLog
v. 0.3beta7 - ?
	Prereleases: 0: Mon Jul 19 2010 12:58 CEST;
	             1: Mon Jul 19 2010 20:37 CEST;
	             2: Wed Jul 21 2010 11:07 CEST
	* Added support for audio profiles
	* Updated libroarrsound
	* Added RPlay protocol to roard
	* Updates of copyright and license headers
	* Added ckport file
	* Updated some tools to get away from old, legacy
	  and security relevant function calls
	* Fixed bug with incorrect number of arguments to roard
	* Added some more ROAR_INFO()s to roard
	* Added tool roarclientpass

v. 0.3beta6 - Thu Jun 24 2010 20:47 CEST
	Prereleases: 0: Fri Jun 11 2010 10:47 CEST;
	             1: Sat Jun 12 2010 11:07 CEST;
	             2: Wed Jun 16 2010 12:52 CEST
	* Added something like select() for VIOs
	* Fixed a lot warnings
	* disabled cmd codecfilter by default
	* Fixed error with roarinterconnect on non-OSS systems
	* Fixed some problems with minimal builds
	* Some Copyright fixes
	* much better VIO support for connections
	* much better OSS emulation
	* Updated win32 port
	* Updated RSound support
	* fixed pos calc in case of PCM->PCM converting (pr1)
	* fixed libroarpulse (pr2)

v. 0.3beta5 - Sun May 02 2010 12:41 CEST
	* added tool roardtmf
	* Support to store server address in X11's root window propertys
	* some small fixes
	* wrote a lot more libroarpulse code, some async API clients work now
	* added compbins for esddsp, artsdsp, audiooss, padsp
	* splited libroarpulse and libroarpulse-simple
	* upgraded some code to use VIO
	* more updates for Debian
	* fixed init script for Debian and Debian like systems
	* fixed some mixer stuff, a lot things work better now
	* added support for notify beeps
	* updated drivers
	* updated ROAR_{ERR,WARN,INFO,DBG}()
	* added syslog support
	* updated test suit
	* done a lot work on libroaross
	* fixed segfault in ALSA plugin
	* Added support for OSS to roarinterconnect(1)
	* Added server side RSound Emulation
	* Added libroarrsound, client side RSound Emulation
	* Added ALSA driver
	* Added RSound driver
	* Added PortAudio driver

v. 0.3beta4 - Sat Feb 06 2010 26:02 CET
	* Applyed patches for Debian
	* updated roarify to load libs of correct version
	* updated list of supported versions (complibs)
	* added header with common units
	* added support for PulseAudio Simple protocol
	* added support to libroaross to emulate DMX4Linux (write only)
	* added DSTR driver to roard
	* updated librarys' soname
	* do not link OpenSSL by default (licence incompatible?)
	* support for dynamic lib loading, plugin interface
	* support to for registering DSTR types
	* moved WinSock things out of main vio.[ch]
	* support auto-tool --prefix*=*

v. 0.3beta3 - Fri Jan 22 2010 22:17 CET
	* added some aRts/NAS/YIFF/PulseAudio tools to compbins
	* added support for comp libs to roar-config
	* added support for uni-directional streams in roarinterconnect
	* write much better support for config, including codec config
	* reduced the TCP overhead produced by speex
	* added support for speex preprocessor
	* get names of complibs from table, do not guess them
	* a lot small fixes
	* moved most of the mixer stuff out of roard into libroardsp
	* added flags immutable and enhance
	* test for signal.h
	* Corrected (DSTR) HTTP Protocol support so servers without headerflush
	  should now work, too.
	* added protocol support for icy
	* Added version numbers to lib names
	* updated manpages
	* wrote OSS emulation
	* fixed test scripts

v. 0.3beta2 - Mon Oct 05 2009 19:48 CEST
	* Added prethru flag
	* Added support to read info about client's network node
	* Ported libroaresd, libroarartsc
	* Fixed some problems with EsounD emulation
	* added support for compatibility binarys
	* added poll() support to libroarsndio
	* reject streams of unsupported codec

v. 0.3beta1 - Sat Sep 26 2009 25:50 CEST
	* incremented nummber of streams per client fron 4 to 8
	* RIFF WAVE codecfilter uses virtual streams
	* another set of changed names of parameters in roarctl
	* Interface to read byteorder of a client from roard
	* implemented a 'state of stream' var
	* cleanup, small fixes
	* added VIO interface to transcoder interface
	* ask codecfilter of parent in case of new child stream
	* added 'subsystem' complex: multi subsystem data on single streams
	* added basic /Radio Data and Trasmitter Control System/.
	* better support for stereo upmixing
	* make audio group a optional thing
	* better support for win32
	* wrote WMM driver
	* better multi protocol support
	* Support for server side EsounD emulation on win32

v. 0.3beta0 - Sun Sep 06 2009 22:42 CEST
	* added new stream directions for MIDI and light control
	* cleanup of client stream struct
	* added basic light control subsystem
	* added DMX512 driver
	* wrote basic MIDI subsystem
	* added flag 'mute'
	* added flag 'mmap'
	* added lib libroarmidi and libroarlight
	* added OpenSLP support
	* rewote fillmixbuer2 code
	* added a roar_conv2() with a more clear API
	* added support to write pidfile
	* added support for encoding 32 bit Vorbis streams
	* added transcoding features in libroardsp
	* added tool 'roarphone'
	* added tool 'roarinterconnect'
	* added simple raw data subsystem
	* added working support for THRU streams
	* changed names of parameters in roarctl
	* added lib libroareio - Extended IO for RoarAudio
	* Update order of default drivers, now it is: oss, ao, esd, null
	* cleanup of old sysio driver interface
	* added server side EsounD emulation
	* added multi homed support to server
	* fixed a long outstanding bug with pa_path_get_filename()

v. 0.2 - Thu May 21 2009 21:00
	* do a some more testings for functions not existing on win32
	* init winsock
	* done a lot fixes in order to get libroar and some clients work on win32
	* done a lot cleanup on configure script

v. 0.2rc2 - Sun May 17 2009 25:44
	* corrected check for GNU peercred
	* added basic logging support to configure
	* fixed plugin configures
	* added and fixed semi-install target on plugins

roarclients/roarclientpass.c
//roarclientpass.c:

/*
 *      Copyright (C) Philipp 'ph3-der-loewe' Schafft - 2010
 *
 *  This file is part of roarclients a part of RoarAudio,
 *  a cross-platform sound system for both, home and professional use.
 *  See README for details.
 *
 *  This file is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License version 3
 *  as published by the Free Software Foundation.
 *
 *  RoarAudio is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this software; see the file COPYING.  If not, write to
 *  the Free Software Foundation, 51 Franklin Street, Fifth Floor,
 *  Boston, MA 02110-1301, USA.
 *
 */

#include <roaraudio.h>

void usage (void) {
 printf("roarclientpass [OPTIONS]...\n");

 printf("\nOptions:\n\n");

 printf("  --server    SERVER    - Set server hostname\n"
        "  --stdin               - Client is on stdin\n"
        "  --stdout              - Client is on stdout\n"
        "  --stdio               - Same as --stdin --stdout\n"
        "  --stderr              - Client is on stderr\n"
        "  --client-fh FH        - Client is on FH\n"
        "  --proto PROTO         - Client uses protocol PROTO (default: RoarAudio)\n"
        "  --byteorder BO        - Client uses byteorder BO (default: network)\n"
        "  --listen              - This is a listen mode connection\n"
        "  --mode MODE           - Set mode of operation: none, listen, connect (default: none)\n"
        "  --bind BIND           - Set host/node/path for mode listen and connect\n"
        "  --port PORT           - Set port for mode listen and connect\n"
//        "  --type TYPE           - Set type for mode listen and connect (default: unknown)\n"
        "  --help                - Show this help\n"
       );

}

#define _BV(x) (1<<(x))
#define F_STDIN  _BV(ROAR_STDIN)
#define F_STDOUT _BV(ROAR_STDOUT)
#define F_STDERR _BV(ROAR_STDERR)

int main (int argc, char * argv[]) {
 struct roar_connection    con;
 struct roar_client        client;
 char * server    = NULL;
 char * k;
 int    i;
 int    clientfh  = -1;
 int    cflags    = 0;
 int    flags     = 0;
 int    proto     = ROAR_PROTO_ROARAUDIO;
 int    byteorder = ROAR_BYTEORDER_NETWORK;
 int    mode      = ROAR_SOCKET_MODE_NONE;
 int    type      = ROAR_SOCKET_TYPE_UNKNOWN;
 char * host      = NULL;
 int    port      = -1;

 for (i = 1; i < argc; i++) {
  k = argv[i];

  if ( !strcmp(k, "--server") || !strcmp(k, "-s") ) {
   server = argv[++i];
  } else if ( !strcmp(k, "--stdin") ) {
   cflags |= F_STDIN;
  } else if ( !strcmp(k, "--stdout") ) {
   cflags |= F_STDOUT;
  } else if ( !strcmp(k, "--stderr") ) {
   cflags |= F_STDERR;
  } else if ( !strcmp(k, "--stdio") ) {
   cflags |= F_STDIN|F_STDOUT;
  } else if ( !strcmp(k, "--client-fh") ) {
   clientfh = atoi(argv[++i]);
  } else if ( !strcmp(k, "--proto") ) {
   proto = roar_str2proto(argv[++i]);
  } else if ( !strcmp(k, "--byteorder") ) {
   byteorder = roar_str2byteorder(argv[++i]);
  } else if ( !strcmp(k, "--listen") ) {
   flags |= ROAR_CLIENTPASS_FLAG_LISTEN;
  } else if ( !strcmp(k, "--mode") ) {
   k = argv[++i];
   if ( !strcasecmp(k, "none") ) {
    mode = ROAR_SOCKET_MODE_NONE;
   } else if ( !strcasecmp(k, "listen") ) {
    mode = ROAR_SOCKET_MODE_LISTEN;
    flags |= ROAR_CLIENTPASS_FLAG_LISTEN;
   } else if ( !strcasecmp(k, "connect") ) {
    mode = ROAR_SOCKET_MODE_CONNECT;
    flags -= ROAR_CLIENTPASS_FLAG_LISTEN;
   } else {
    ROAR_ERR("unknown mode: %s", k);
    return 1;
   }
  } else if ( !strcmp(k, "--bind") ) {
   host = argv[++i];
  } else if ( !strcmp(k, "--port") ) {
   port = atoi(argv[++i]);
  } else if ( !strcmp(k, "--help") || !strcmp(k, "-h") ) {
   usage();
   return 0;
  } else {
   ROAR_ERR("unknown argument: %s", k);
   usage();
   return 1;
  }
 }

 if ( cflags & F_STDERR ) {
#ifdef ROAR_HAVE_SYSLOG
  roar_debug_set_stderr_mode(ROAR_DEBUG_MODE_SYSLOG);
#else
  roar_debug_set_stderr_fh(-1);
#endif
 } else {
  roar_debug_set_stderr_fh(ROAR_STDERR);
 }

 if ( mode != ROAR_SOCKET_MODE_NONE ) {
  if ( clientfh != -1 ) {
   ROAR_ERR("Too may socket types given");
   return 30;
  }

  clientfh = roar_socket_open(mode, type, host, port);

  if ( clientfh == -1 ) {
   ROAR_ERR("Unabled to open socket");
   return 31;
  }
 }

 if ( clientfh == -1 ) {
  if ( cflags & F_STDIN ) {
   clientfh = ROAR_STDIN;
  } else if ( cflags & F_STDOUT ) {
   clientfh = ROAR_STDOUT;
  } else if ( cflags & F_STDERR ) {
   clientfh = ROAR_STDERR;
  } else {
   ROAR_ERR("No client socket given");
   return 32;
  }
 }

 roar_client_new(&client);
 roar_client_set_fh(&client, clientfh);
 roar_client_set_proto(&client, proto, byteorder);

 if ( roar_simple_connect(&con, server, "roarclientpass") == -1 ) {
  ROAR_ERR("Can not connect to server");
  return 10;
 }

 if ( roar_client_pass(&con, &client, flags) == -1 ) {
  ROAR_ERR("Can not pass client fh to server");
  roar_disconnect(&con);
  return 20;
 }

 roar_disconnect(&con);

 return 0;
}

//ll
